components/style/values/specified/length.rs

```diff
@@ -1117,10 +1117,10 @@ pub type LengthOrNumber = Either<Length, Number>;
 impl LengthOrNumber {
     /// Parse a non-negative LengthOrNumber.
     pub fn parse_non_negative(input: &mut Parser) -> Result<Self, ()> {
-        if let Ok(v) = input.try(Length::parse_non_negative) {
-            Ok(Either::First(v))
+        if let Ok(v) = input.try(Number::parse_non_negative) {
```

Comment:
Nit: You should probably drop a comment here as to why we're parsing the number first. Something like,

```
Both `Length` and `Number` can parse "0", but for cases like `LengthOrNumber`
we want "0" to be parsed and serialized as a number instead of "0px" (which 
matches the behavior of all browsers)
```